### PR TITLE
Visualization port

### PR DESCRIPTION
**Pull request by nzlz**
### Description
Porting `moveit_rviz_plugin_render_tools` inside `moveit_ros/visualization`. The goal of the PR is to save/backup the branch in the Acutronic repo instead of my personal one. 

- Compiling: YES
- Working: not tested

### CODE REVIEW

File: moveit_ros/visualization/trajectory_rviz_plugin_description.xml
Comment (ahcorde) [inline moveit_ros/visualization/trajectory_rviz_plugin_description.xml]: you should remove `lib`

File: moveit_ros/visualization/CMakeLists.txt
Comment (ahcorde) [inline moveit_ros/visualization/CMakeLists.txt:106]: you should replace this line with 
```
pluginlib_export_plugin_description_file(rviz_common plugins_description.xml)
```

Have a look here https://github.com/AcutronicRobotics/rviz_hrim_plugins/blob/master/CMakeLists.txt

File: moveit_ros/visualization/planning_scene_rviz_plugin_description.xml
Comment (ahcorde) [inline moveit_ros/visualization/planning_scene_rviz_plugin_description.xml]: you should remove `lib`

File: moveit_ros/visualization/robot_state_rviz_plugin_description.xml
Comment (ahcorde) [inline moveit_ros/visualization/robot_state_rviz_plugin_description.xml]: you should remove `lib`

**Review by ahcorde — CHANGES_REQUESTED** [review]

File: moveit_ros/visualization/rviz_plugin_render_tools/include/moveit/rviz_plugin_render_tools/trajectory_visualization.hpp
Comment (anasarrak) [inline moveit_ros/visualization/rviz_plugin_render_tools/include/moveit/rviz_plugin_render_tools/trajectory_visualization.hpp]: Why remove the const and the reference (&)?

File: moveit_ros/visualization/rviz_plugin_render_tools/include/moveit/rviz_plugin_render_tools/trajectory_visualization.hpp
Comment (anasarrak) [inline moveit_ros/visualization/rviz_plugin_render_tools/include/moveit/rviz_plugin_render_tools/trajectory_visualization.hpp]: Look at this https://answers.ros.org/question/212857/what-is-constptr/

File: moveit_ros/visualization/rviz_plugin_render_tools/src/trajectory_visualization.cpp
Comment (anasarrak) [inline moveit_ros/visualization/rviz_plugin_render_tools/src/trajectory_visualization.cpp]: Rever back the const

File: moveit_ros/visualization/motion_planning_rviz_plugin/include/moveit/motion_planning_rviz_plugin/motion_planning_frame.hpp
Comment (anasarrak) [inline moveit_ros/visualization/motion_planning_rviz_plugin/include/moveit/motion_planning_rviz_plugin/motion_planning_frame.hpp]: Is it this working? The returning time from the clock().now() should be a rclcpp::Time
http://docs.ros2.org/bouncy/api/rclcpp/classrclcpp_1_1_clock.html#a6375208429c40bd9d6e84b2b32557182, am I wrong?

File: moveit_ros/visualization/robot_state_rviz_plugin/src/robot_state_display.cpp
Comment (anasarrak) [inline moveit_ros/visualization/robot_state_rviz_plugin/src/robot_state_display.cpp]: Use rclcpp::Clock().now() as a replacement for ros::time::now(), to replace the ros::time you can just use rclcpp::Time :) .  Please, fix the previous ones.

Comment (nzlz) [reply moveit_ros/visualization/robot_state_rviz_plugin/src/robot_state_display.cpp]: This is `Time(0)`, which intends to get the latest available time available AFAIK. But looking at the function and I guess getting current time matches better the expected behavior. Maybe you propose to use rclcpp::Time::now ? instead of Clock.

File: moveit_ros/visualization/rviz_plugin_render_tools/include/moveit/rviz_plugin_render_tools/trajectory_visualization.hpp
Comment (nzlz) [reply moveit_ros/visualization/rviz_plugin_render_tools/include/moveit/rviz_plugin_render_tools/trajectory_visualization.hpp]: removing & because using SharedPtr is a common practice in ROS2. regarding the **const**, I find strange to use it with a node pointer, but will research this.

Note that the original moveit1code is `const ros::NodeHandle& update_nh`

File: moveit_ros/visualization/rviz_plugin_render_tools/include/moveit/rviz_plugin_render_tools/trajectory_visualization.hpp
Comment (nzlz) [reply moveit_ros/visualization/rviz_plugin_render_tools/include/moveit/rviz_plugin_render_tools/trajectory_visualization.hpp]: ConstPr was throwing warnings regarding a current or future deprecated state, not sure. Maybe we should replace `ConstPtr` for `const X SharedPtr` ?

File: moveit_ros/visualization/motion_planning_rviz_plugin/include/moveit/motion_planning_rviz_plugin/motion_planning_frame.hpp
Comment (nzlz) [reply moveit_ros/visualization/motion_planning_rviz_plugin/include/moveit/motion_planning_rviz_plugin/motion_planning_frame.hpp]: I think you are right. Is it okay if I replace it with `auto`?

File: moveit_ros/visualization/robot_state_rviz_plugin/src/robot_state_display.cpp
Comment (anasarrak) [reply moveit_ros/visualization/robot_state_rviz_plugin/src/robot_state_display.cpp]: The returning value from a Clock.now() is a Time yes, but is not the same as declaring it with rclcpp::Time(0). With time(0) you are just passing the time as 0, and with Clock().now() you are just getting the rcl time each second/millisecond. For our previous changes we just replace this kind of changes with the rclcpp::time(), as an example you can take this:

https://github.com/AcutronicRobotics/moveit2/commit/022da7cbbdb161df80e1fb8a9454f3c03f749ce9#diff-7e8d20a58c82e13dc1c7c8369618bcfbR153

File: moveit_ros/visualization/rviz_plugin_render_tools/include/moveit/rviz_plugin_render_tools/trajectory_visualization.hpp
Comment (anasarrak) [reply moveit_ros/visualization/rviz_plugin_render_tools/include/moveit/rviz_plugin_render_tools/trajectory_visualization.hpp]: Williams says on the link that I sent you that a replacement for a Ptr shoud be as following:

``
std_msgs::msg::String::Ptr
`` to `` std::shared_ptr<std_msgs::msg::String>`` or ``std_msgs::msg::String::SharedPtr``

and with constPtr
`` std::shared_ptr<const std_msgs::msg::String>`` It says that using the const before the msg sometimes is needed because for *std_msgs::string* it might not make much difference, but for example for a *sensor_msgs::PointCloud2* it might.

File: moveit_ros/visualization/rviz_plugin_render_tools/include/moveit/rviz_plugin_render_tools/trajectory_visualization.hpp
Comment (anasarrak) [reply moveit_ros/visualization/rviz_plugin_render_tools/include/moveit/rviz_plugin_render_tools/trajectory_visualization.hpp]: We usually leave it as a const, if it's already a const

File: moveit_ros/visualization/motion_planning_rviz_plugin/include/moveit/motion_planning_rviz_plugin/motion_planning_frame.hpp
Comment (anasarrak) [reply moveit_ros/visualization/motion_planning_rviz_plugin/include/moveit/motion_planning_rviz_plugin/motion_planning_frame.hpp]: Yes, we usually use *auto* for that

File: moveit_ros/visualization/rviz_plugin_render_tools/src/trajectory_visualization.cpp
Comment (nzlz) [reply moveit_ros/visualization/rviz_plugin_render_tools/src/trajectory_visualization.cpp]: 0fc4f6629d8e7ffadaa13200233f1c483bb1f241

File: moveit_ros/visualization/motion_planning_rviz_plugin/include/moveit/motion_planning_rviz_plugin/motion_planning_frame.hpp
Comment (nzlz) [reply moveit_ros/visualization/motion_planning_rviz_plugin/include/moveit/motion_planning_rviz_plugin/motion_planning_frame.hpp]: 0fc4f6629d8e7ffadaa13200233f1c483bb1f241

File: moveit_ros/visualization/robot_state_rviz_plugin/src/robot_state_display.cpp
Comment (nzlz) [reply moveit_ros/visualization/robot_state_rviz_plugin/src/robot_state_display.cpp]: 0fc4f6629d8e7ffadaa13200233f1c483bb1f241

File: moveit_ros/visualization/rviz_plugin_render_tools/include/moveit/rviz_plugin_render_tools/trajectory_visualization.hpp
Comment (nzlz) [reply moveit_ros/visualization/rviz_plugin_render_tools/include/moveit/rviz_plugin_render_tools/trajectory_visualization.hpp]: Added const to the SharedPtr
0fc4f6629d8e7ffadaa13200233f1c483bb1f241

**Comment by ahcorde** [comment]
I resolved the conflicts to be able to merge it into master. @nzlz can you check that I didn't remove something relevant?

File: moveit_ros/visualization/rviz_plugin_render_tools/include/moveit/rviz_plugin_render_tools/trajectory_visualization.hpp
Comment (nzlz) [reply moveit_ros/visualization/rviz_plugin_render_tools/include/moveit/rviz_plugin_render_tools/trajectory_visualization.hpp]: We are going with `ConstSharedPtr`, info here: https://design.ros2.org/articles/generated_interfaces_cpp.html

Resolving conversation.

File: moveit_ros/visualization/CMakeLists.txt
Comment (nzlz) [reply moveit_ros/visualization/CMakeLists.txt:106]: Used this for the robot_state_rviz_plugin_description, will have to do the same with others once ported. 
https://github.com/nzlz/moveit2/blob/3eeba3abb3d9a5c8c3e61ff0c02308260de50e5c/moveit_ros/visualization/CMakeLists.txt#L107
`pluginlib_export_plugin_description_file(rviz_common robot_state_rviz_plugin_description.xml)`

**Comment by nzlz** [comment]
Compiling fine (only warnings) after latest move_group fixes. https://github.com/AcutronicRobotics/moveit2/pull/103#issuecomment-499345678 .

Ready to merge in the visualization branch IMHO.

**Comment by anasarrak** [comment]
It seems like *moveit_ros_visualization* is not passing the CI 
``` 
Summary: 35 packages finished [18min 21s]
  1 package failed: moveit_ros_visualization
  1 package aborted: moveit_ros_manipulation
```

**Comment by LanderU** [comment]
I'm working to fix the CI to have this and other PR merged ASAP.

**Comment by nzlz** [comment]
The port of `motion_planning_rviz_plugin` will be long and complex. You can take a look at the unported dependencies here: 
 https://github.com/nzlz/moveit2/tree/visualization/moveit_ros/visualization#motion_planning_rviz_plugin-unported-dependencies

I'll dive into these issues and see the feasibility of the migration and plan how to proceed, but input is welcomed.

**Comment by LanderU** [comment]
CI passed!

**Comment by nzlz** [comment]
Great! Can we merge and start working in the acutronic repo (visulization branch)?

**Comment by LanderU** [comment]
@ahcorde @anasarrak, review it this again.

File: moveit_ros/visualization/CMakeLists.txt
Comment (anasarrak) [inline moveit_ros/visualization/CMakeLists.txt:31]: Lets add a TODO, so we dont forget to enable it once everything is merged

File: moveit_ros/visualization/motion_planning_rviz_plugin/CMakeLists.txt
Comment (anasarrak) [inline moveit_ros/visualization/motion_planning_rviz_plugin/CMakeLists.txt:30]:
```suggestion
  ${rclcpp_LIBRARIES} ${rviz_DEFAULT_PLUGIN_LIBRARIES} ${QT_LIBRARIES} ${Boost_LIBRARIES})
```

File: moveit_ros/visualization/motion_planning_rviz_plugin/include/moveit/motion_planning_rviz_plugin/motion_planning_display.hpp
Comment (anasarrak) [inline moveit_ros/visualization/motion_planning_rviz_plugin/include/moveit/motion_planning_rviz_plugin/motion_planning_display.hpp]: Why are we changing the namespace?

File: moveit_ros/visualization/motion_planning_rviz_plugin/include/moveit/motion_planning_rviz_plugin/motion_planning_frame.hpp
Comment (anasarrak) [inline moveit_ros/visualization/motion_planning_rviz_plugin/include/moveit/motion_planning_rviz_plugin/motion_planning_frame.hpp]: Replaceable for rclcpp_action, can we remove the commented line?

File: moveit_ros/visualization/motion_planning_rviz_plugin/include/moveit/motion_planning_rviz_plugin/motion_planning_frame.hpp
Comment (anasarrak) [inline moveit_ros/visualization/motion_planning_rviz_plugin/include/moveit/motion_planning_rviz_plugin/motion_planning_frame.hpp]: Same as https://github.com/AcutronicRobotics/moveit2/pull/101/files#r292771581

File: moveit_ros/visualization/motion_planning_rviz_plugin/include/moveit/motion_planning_rviz_plugin/motion_planning_display.hpp
Comment (nzlz) [reply moveit_ros/visualization/motion_planning_rviz_plugin/include/moveit/motion_planning_rviz_plugin/motion_planning_display.hpp]: Rviz2 plugins use that namespace because most rviz core funcionalities and classes are now in rviz_common.

**Comment by nzlz** [comment]
@anasarrak the motion planning plugin you are referring to is not ported to ros2. Reason : https://github.com/nzlz/moveit2/tree/visualization/moveit_ros/visualization#motion_planning_rviz_plugin-unported-dependencies

Do you want to fix those before merge anyway?

File: moveit_ros/visualization/motion_planning_rviz_plugin/src/motion_planning_display.cpp
Comment (anasarrak) [inline moveit_ros/visualization/motion_planning_rviz_plugin/src/motion_planning_display.cpp]: You need a Logger name here, you have to declare it as 
``rclcpp::Logger LOGGER_MOTION_PLANNING_RVIZ_PLUGIN = rclcpp::get_logger("motion_planning_rviz_plugin").get_child("motion_planning_display");`` for example, or if it is already declared a node, you can just use it. 
```suggestion
    RCLCPP_ERROR(node_->get_logger(),"Group [%s] not found in the robot model.", group.c_str());
```
or
```suggestion
    RCLCPP_ERROR(LOGGER_MOTION_PLANNING_RVIZ_PLUGIN,"Group [%s] not found in the robot model.", group.c_str());
```

File: moveit_ros/visualization/motion_planning_rviz_plugin/src/motion_planning_frame.cpp
Comment (anasarrak) [inline moveit_ros/visualization/motion_planning_rviz_plugin/src/motion_planning_frame.cpp]: Enable this again, use logger name as specified https://github.com/AcutronicRobotics/moveit2/pull/101#discussion_r292774419

**Comment by anasarrak** [comment]
@nzlz  and I have discussed about the **motion_planning_rviz_plugin**, taking into account that the plugin is not ported and needs further work, we haved decided to leave it as master, and create a branch with the current work, so the changes are not lost. There are many changes that needs to be reviewed and right now because of https://github.com/AcutronicRobotics/moveit2/pull/101#issuecomment-501151147 we cannot reproduce it. What's your opinion about that @ahcorde ?

motion_planning_rviz_plugin is the only one that is not ported, the remaining ones should be ported.

File: moveit_ros/visualization/planning_scene_rviz_plugin/CMakeLists.txt
Comment (anasarrak) [inline moveit_ros/visualization/planning_scene_rviz_plugin/CMakeLists.txt:13]: This line has to be reverted back as
```suggestion
install(DIRECTORY include/ DESTINATION include)
```

File: moveit_ros/visualization/rviz_plugin_render_tools/src/trajectory_visualization.cpp
Comment (anasarrak) [inline moveit_ros/visualization/rviz_plugin_render_tools/src/trajectory_visualization.cpp]: We normally replace this as RCLCPP_ERROR

File: moveit_ros/visualization/rviz_plugin_render_tools/src/trajectory_visualization.cpp
Comment (anasarrak) [inline moveit_ros/visualization/rviz_plugin_render_tools/src/trajectory_visualization.cpp]: It's prefered to use RCLCPP_INFO as we discussed a while ago with williams

File: moveit_ros/visualization/trajectory_rviz_plugin/CMakeLists.txt
Comment (anasarrak) [inline moveit_ros/visualization/trajectory_rviz_plugin/CMakeLists.txt:27]: I think that this install should be reverted back

**Comment by anasarrak** [comment]
Lets wait the CI, if it passes we can merge this

File: moveit_ros/visualization/CMakeLists.txt
Comment (ahcorde) [inline moveit_ros/visualization/CMakeLists.txt]: remove this TODO

File: moveit_ros/visualization/CMakeLists.txt
Comment (ahcorde) [inline moveit_ros/visualization/CMakeLists.txt]: the first argument should be `rviz_common` like this:

```
pluginlib_export_plugin_description_file(rviz_common robot_state_rviz_plugin_description.xml)
pluginlib_export_plugin_description_file(rviz_common planning_scene_rviz_plugin_description.xml)
pluginlib_export_plugin_description_file(rviz_common trajectory_rviz_plugin_description.xml)
```

File: moveit_ros/visualization/CMakeLists.txt
Comment (ahcorde) [inline moveit_ros/visualization/CMakeLists.txt]: ?

File: moveit_ros/visualization/package.xml
Comment (ahcorde) [inline moveit_ros/visualization/package.xml]: why these lines are commented?

**Review by ahcorde — CHANGES_REQUESTED** [review]

File: moveit_ros/visualization/package.xml
Comment (nzlz) [reply moveit_ros/visualization/package.xml]: Ill remove them. They are from the ros1 repo.

**Comment by nzlz** [comment]
Updated.

**Comment by anasarrak** [comment]
I cherry-picked the changes for the CI

**Comment by ahcorde** [comment]
Let's wait for the CI again

**Comment by nzlz** [comment]
Ping, CI passing.

**Comment by vmayoral** [comment]
What's up with this guys?

**Comment by anasarrak** [comment]
This can be merged, waiting for @ahcorde's approval

**Pull request merged**
